### PR TITLE
Mail command

### PR DESCRIPTION
**Pull request by markediez**
Output mail forwarding information of a user

### CODE REVIEW

File: commands/mail.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+      response = get_information message.split(" ")[1]
+
+      # Format information
+      format_information response
+    end
+
+    # Returns formatted string of the user's mail forwarding information
+    # @param data - an array of Nokogiri objects
+    def format_information(data)
+      return "No user found" unless !data.empty?
```
Comment (cthielen) [inline commands/mail.rb]: if data.empty? is better than unless !data.empty?

File: commands/mail.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    # @return - an array of Nokogiri objects
+    def get_information(login_id)
+      # Success! Load the page and use the CAS credentials
+      agent = Mechanize.new
+
+      # Certain web servers check the user agent string for compatibility
+      agent.user_agent_alias = 'Mac Firefox'
+
+      # Run Mail ID lookup
+      page = agent.get('https://iet.ucdavis.edu/itprovider')
```
Comment (cthielen) [inline commands/mail.rb]: What is agent.get() fails?

File: commands/mail.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+      # Find form TODO: try to see if we can find it dynamically?
+      mail_id_form = page.forms[2]
```
Comment (cthielen) [inline commands/mail.rb]: What if the page changes and this isn't the right form or the form doesn't exist? Does the form have a name or ID we can confirm?

File: commands/mail.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+      # Fill form
+      mail_id_form.fields[0].value = login_id
+
+      # Submit Form
+      page = agent.submit(mail_id_form, mail_id_form.buttons.first)
+
+      unless page.code == "200"
+        return "Could not complete the task due to " + page.code
```
Comment (cthielen) [inline commands/mail.rb]: "Could not fetch Mail ID information due to error while reading webpage. Received code: " or something is clearer to the user.

File: commands/mail.rb [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+      end
+
+      return page.css("div.form-item.form-type-item")
```
Comment (cthielen) [inline commands/mail.rb]: What if this item doesn't exist?

File: commands/mail.rb [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  class Mail
+    TITLE = "Mail ID"
+    REGEX = /^!mailid\s+\S+\s*$/
+    COMMAND = "!mailid <Login ID>"
+    DESCRIPTION = "Shows the mail forwarding information of a user e.g. `!mailid cmthielen`"
+
+    # Run MUST return a string
+    def run(message, channel)
+      # Get information
+      response = get_information message.split(" ")[1]
```
Comment (cthielen) [inline commands/mail.rb]: What if get_information() returned an error code?

File: Gemfile [hunk 7]
```diff
@@ -16,3 +16,6 @@ gem 'faraday-http-cache' # (provides etag caching in octokit)
 
 # For cron tasks
 gem 'whenever', :require => false
+
+# For crawling
```
Comment (cthielen) [inline Gemfile]: Note that it is for the Mail ID command specifically.

File: commands/mail.rb [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+      value = data[6].text.split(" ").pop
+      response.push "*SMTP* #{value}"
+
+      return response.join("\n")
+    end
+
+    # Retrieves mail forwarding infomration from iet.ucdavis.edu
+    # @param login_id - Kerberos login id to query
+    # @return - an array of Nokogiri objects
+    def get_information(login_id)
```
Comment (cthielen) [inline commands/mail.rb]: Rename to fetch_mail_settings()

File: commands/mail.rb [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+      # Get information
+      response = get_information message.split(" ")[1]
+
+      # Format information
+      format_information response
+    end
+
+    # Returns formatted string of the user's mail forwarding information
+    # @param data - an array of Nokogiri objects
+    def format_information(data)
```
Comment (cthielen) [inline commands/mail.rb]: Rename to format_response()

File: commands/mail.rb (hunk 3 above)
Comment (markediez) [inline commands/mail.rb]: Unfortunately it doesn't have a name or id, the only specific attribute it has is a class

File: commands/mail.rb [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+        # Submit Form
+        page = agent.submit(mail_id_form, mail_id_form.buttons.first)
+      rescue SocketError
+        $logger.error "Could not fetch Mail ID information due to error loading the webpage. Did the link change?"
+        return "Could not fetch Mail ID information due to error loading the webpage"
+      else
+        unless page.code == "200"
+          $logger.warn "Could not fetch Mail ID information due to error while reading webpage. Received code:  " + page.code
+          return "Could not fetch Mail ID information due to error while reading webpage. Received code:  " + page.code
+        end
```
Comment (cthielen) [inline commands/mail.rb]: Are you missing an 'end' here? The 'else' matches the begin, so lines 95-101 only happen when there's an exception. They look like regular code though.

File: commands/mail.rb [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    REGEX = /^!mailid\s+\S+\s*$/
+    COMMAND = "!mailid <Login ID>"
+    DESCRIPTION = "Shows the mail forwarding information of a user e.g. `!mailid cmthielen`"
+
+    # Run MUST return a string
+    def run(message, channel)
+      # Get information
+      response = fetch_mail_settings message.split(" ")[1]
+
+      return response if response.class == String
```
Comment (cthielen) [inline commands/mail.rb]: Clearer to say "if response.class != Nokogiri objects".

File: commands/mail.rb [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+        unless page.code == "200"
+          $logger.warn "Could not fetch Mail ID information due to error while reading webpage. Received code:  " + page.code
+          return "Could not fetch Mail ID information due to error while reading webpage. Received code:  " + page.code
+        end
+
+        forwarding_info = page.css("div.form-item.form-type-item")
+        unless forwarding_info.empty?
+          return forwarding_info
+        else
+          $logger.error "Mail ID information not found. Did the page format change?"
```
Comment (cthielen) [inline commands/mail.rb]: How about: "Mail ID information not found. Page layout not recognized."

"Did the page format change?" sounds like a useful thing to see if you're a developer but the IT help desk folks won't know what it's referring to. They won't necessarily know that this command operates via a crawl. Keep the users in mind.

**Pull request merged**